---

title: Liftover module


keywords: fastai
sidebar: home_sidebar

summary: "Liftover genodata and sumstat"
description: "Liftover genodata and sumstat"
nb_path: "nbs/02_Liftover.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/02_Liftover.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="Liftover" class="doc_header"><code>class</code> <code>Liftover</code><a href="https://github.com/changebio/LDtools/tree/master/LDtools/liftover.py#L10" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Liftover</code>(<strong><code>fr</code></strong>=<em><code>'hg19'</code></em>, <strong><code>to</code></strong>=<em><code>'hg38'</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>424</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">LDtools.genodata</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">LDtools.sumstat</span> <span class="kn">import</span> <span class="n">Sumstat</span>
<span class="kn">from</span> <span class="nn">LDtools.liftover</span> <span class="kn">import</span> <span class="n">Liftover</span>
<span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">fr</span><span class="o">=</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span><span class="n">to</span><span class="o">=</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span><span class="n">remove_missing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">rename</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="n">lf</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="n">fr</span><span class="p">,</span><span class="n">to</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;liftover from &quot;</span> <span class="o">+</span> <span class="n">fr</span> <span class="o">+</span><span class="s2">&quot;to&quot;</span> <span class="o">+</span><span class="n">to</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Removing SNPs failed to liftover is&quot;</span><span class="p">,</span> <span class="n">remove_missing</span><span class="p">)</span>
    <span class="c1">#file type detection, sumstats, plink, vcf,gvcf, &gt;&gt;&gt;future bgen</span>
    <span class="n">input_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">input_suffixes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">input_path</span><span class="o">.</span><span class="n">suffixes</span><span class="p">)</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">input_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span> <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The file is not exist:&quot;</span><span class="p">,</span> <span class="n">input_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">input_path</span><span class="o">.</span><span class="n">suffix</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;.bim&#39;</span><span class="p">,</span><span class="s1">&#39;.bed&#39;</span><span class="p">,</span><span class="s1">&#39;.fam&#39;</span><span class="p">]:</span>
        <span class="n">geno</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">input_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">)))</span>
        <span class="n">new_bim</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">bim_liftover</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">bim</span><span class="p">)</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">new_bim</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">remove_missing</span><span class="p">:</span>
            <span class="n">geno</span><span class="o">.</span><span class="n">bim</span> <span class="o">=</span> <span class="n">new_bim</span>
            <span class="n">geno</span><span class="o">.</span><span class="n">extractbyidx</span><span class="p">(</span><span class="o">~</span><span class="n">idx</span><span class="p">)</span>
            <span class="n">geno</span><span class="o">.</span><span class="n">export_plink</span><span class="p">(</span><span class="n">output_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total number SNPs &quot;</span><span class="p">,</span><span class="n">new_bim</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;. Removing SNPs failed to liftover &quot;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">idx</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">write_bim</span><span class="p">(</span><span class="n">output_path</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bim&#39;</span><span class="p">),</span><span class="n">new_bim</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total number SNPs &quot;</span><span class="p">,</span><span class="n">new_bim</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;. The number of SNPs failed to liftover &quot;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">idx</span><span class="p">),</span><span class="s2">&quot;. Their chr and pos is replaced with 0, 0&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">input_suffixes</span><span class="o">.</span><span class="n">intersection</span><span class="p">([</span><span class="s1">&#39;.gvcf&#39;</span><span class="p">,</span><span class="s1">&#39;.vcf&#39;</span><span class="p">]))</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">lf</span><span class="o">.</span><span class="n">vcf_liftover</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">remove_missing</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;This file is considered as sumstat format file&quot;</span><span class="p">)</span>
        <span class="n">sums</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">rename</span><span class="o">=</span><span class="n">rename</span><span class="p">)</span>
        <span class="n">new_sums</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">sumstat_liftover</span><span class="p">(</span><span class="n">sums</span><span class="o">.</span><span class="n">ss</span><span class="p">,</span><span class="n">rename</span><span class="p">)</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">new_sums</span><span class="o">.</span><span class="n">CHR</span> <span class="o">==</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">remove_missing</span><span class="p">:</span>
            <span class="n">new_sums</span><span class="p">[</span><span class="o">~</span><span class="n">idx</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total number SNPs &quot;</span><span class="p">,</span><span class="n">new_sums</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;. Removing SNPs failed to liftover &quot;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">idx</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_sums</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total number SNPs &quot;</span><span class="p">,</span><span class="n">new_sums</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="s2">&quot;. The number of SNPs failed to liftover &quot;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">idx</span><span class="p">),</span><span class="s2">&quot;. Their chr and pos is replaced with 0, 0&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;data/GH.AR.SAD.P1.001.0_X3547_S42_1180478_GVCF.hard-filtered.gvcf.gz&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span><span class="o">.</span><span class="n">suffix</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;bim&#39;</span><span class="p">,</span><span class="s1">&#39;bed&#39;</span><span class="p">,</span><span class="s1">&#39;fam&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>False</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s1">&#39;.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>Path(&#39;test.bed&#39;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="1.Test-liftover-vcf-and-gvcf">1.Test liftover vcf and gvcf<a class="anchor-link" href="#1.Test-liftover-vcf-and-gvcf"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span><span class="s1">&#39;hg38&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">chainmap</span><span class="p">[</span><span class="mi">22</span><span class="p">][</span><span class="mi">50549067</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">vcf</span> <span class="o">=</span><span class="s1">&#39;data/GH.AR.SAD.P1.001.0_X3547_S42_1180478_GVCF.hard-filtered.gvcf.gz&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">main</span><span class="p">(</span><span class="n">vcf</span><span class="p">,</span><span class="s1">&#39;data/new_hg19_hg38_test.gvcf.gz&#39;</span><span class="p">,</span><span class="n">remove_missing</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>liftover from hg19tohg38
Removing SNPs failed to liftover is True
Total number SNPs  93816694 . The number of SNPs failed to liftover  59995
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">region_liftover</span><span class="p">([</span><span class="mi">5</span><span class="p">,</span><span class="mi">272741</span><span class="p">,</span><span class="mi">1213528</span><span class="o">-</span><span class="mi">900000</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(5, 272626, 313413)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">vcf_liftover</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="mi">59995</span><span class="o">/</span><span class="mi">93816694</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>0.0006394917305442463</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="2.Test-liftover-plink">2.Test liftover plink<a class="anchor-link" href="#2.Test-liftover-plink"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="mi">1</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">LDtools.genodata</span> <span class="kn">import</span> <span class="n">Genodata</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/alzheimers-family/linkage_files/geno/full_sample/bfiles/full_sample.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">OSError</span>                                   Traceback (most recent call last)
<span class="ansi-green-fg">/tmp/2206559.1.plot.q/ipykernel_14505/1401600537.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">----&gt; 1</span><span class="ansi-red-fg"> </span>geno <span class="ansi-blue-fg">=</span> Genodata<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#39;/mnt/mfs/statgen/alzheimers-family/linkage_files/geno/full_sample/bfiles/full_sample.bed&#39;</span><span class="ansi-blue-fg">)</span>

<span class="ansi-green-fg">/mnt/mfs/statgen/yin/Github/LDtools/LDtools/genodata.py</span> in <span class="ansi-cyan-fg">__init__</span><span class="ansi-blue-fg">(self, geno_path, sample_path)</span>
<span class="ansi-green-intense-fg ansi-bold">    108</span> <span class="ansi-green-fg">class</span> Genodata<span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">    109</span>     <span class="ansi-green-fg">def</span> __init__<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span>geno_path<span class="ansi-blue-fg">,</span>sample_path<span class="ansi-blue-fg">=</span><span class="ansi-green-fg">None</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 110</span><span class="ansi-red-fg">         </span>self<span class="ansi-blue-fg">.</span>bim<span class="ansi-blue-fg">,</span>self<span class="ansi-blue-fg">.</span>fam<span class="ansi-blue-fg">,</span>self<span class="ansi-blue-fg">.</span>bed <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>read_geno<span class="ansi-blue-fg">(</span>geno_path<span class="ansi-blue-fg">,</span>sample_path<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    111</span> 
<span class="ansi-green-intense-fg ansi-bold">    112</span>     <span class="ansi-green-fg">def</span> __repr__<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">/mnt/mfs/statgen/yin/Github/LDtools/LDtools/genodata.py</span> in <span class="ansi-cyan-fg">read_geno</span><span class="ansi-blue-fg">(self, geno_file, sample_file)</span>
<span class="ansi-green-intense-fg ansi-bold">    115</span>     <span class="ansi-green-fg">def</span> read_geno<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span>geno_file<span class="ansi-blue-fg">,</span>sample_file<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">    116</span>         <span class="ansi-green-fg">if</span> geno_file<span class="ansi-blue-fg">.</span>endswith<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#39;.bed&#39;</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 117</span><span class="ansi-red-fg">             </span>bim<span class="ansi-blue-fg">,</span>fam<span class="ansi-blue-fg">,</span>bed <span class="ansi-blue-fg">=</span>  read_plink<span class="ansi-blue-fg">(</span>geno_file<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">:</span><span class="ansi-blue-fg">-</span><span class="ansi-cyan-fg">4</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">,</span> verbose<span class="ansi-blue-fg">=</span><span class="ansi-green-fg">False</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    118</span>             bim<span class="ansi-blue-fg">.</span>snp <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">&#39;chr&#39;</span><span class="ansi-blue-fg">+</span>bim<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#39;chrom&#39;</span><span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">&#39;pos&#39;</span><span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">&#39;a0&#39;</span><span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">&#39;a1&#39;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">.</span>astype<span class="ansi-blue-fg">(</span>str<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">.</span>agg<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#39;:&#39;</span><span class="ansi-blue-fg">.</span>join<span class="ansi-blue-fg">,</span> axis<span class="ansi-blue-fg">=</span><span class="ansi-cyan-fg">1</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    119</span>         <span class="ansi-green-fg">elif</span> geno_file<span class="ansi-blue-fg">.</span>endswith<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#39;.bgen&#39;</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/pandas_plink/_read.py</span> in <span class="ansi-cyan-fg">read_plink</span><span class="ansi-blue-fg">(file_prefix, verbose)</span>
<span class="ansi-green-intense-fg ansi-bold">    120</span> 
<span class="ansi-green-intense-fg ansi-bold">    121</span>     ref <span class="ansi-blue-fg">=</span> Allele<span class="ansi-blue-fg">.</span>a1
<span class="ansi-green-fg">--&gt; 122</span><span class="ansi-red-fg">     bed = _read_file(
</span><span class="ansi-green-intense-fg ansi-bold">    123</span>         fn<span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">    124</span>         <span class="ansi-green-fg">lambda</span> f<span class="ansi-blue-fg">:</span> _read_bed<span class="ansi-blue-fg">(</span>f<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;bed&#34;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">,</span> nsamples<span class="ansi-blue-fg">,</span> nmarkers<span class="ansi-blue-fg">[</span>f<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;bed&#34;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">,</span> ref<span class="ansi-blue-fg">,</span> Chunk<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">.</span>T<span class="ansi-blue-fg">,</span>

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/pandas_plink/_read.py</span> in <span class="ansi-cyan-fg">_read_file</span><span class="ansi-blue-fg">(fn, read_func, pbar)</span>
<span class="ansi-green-intense-fg ansi-bold">    329</span>     data <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">]</span>
<span class="ansi-green-intense-fg ansi-bold">    330</span>     <span class="ansi-green-fg">for</span> f <span class="ansi-green-fg">in</span> fn<span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 331</span><span class="ansi-red-fg">         </span>data<span class="ansi-blue-fg">.</span>append<span class="ansi-blue-fg">(</span>read_func<span class="ansi-blue-fg">(</span>f<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    332</span>         pbar<span class="ansi-blue-fg">.</span>update<span class="ansi-blue-fg">(</span><span class="ansi-cyan-fg">1</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    333</span>     <span class="ansi-green-fg">return</span> data

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/pandas_plink/_read.py</span> in <span class="ansi-cyan-fg">&lt;lambda&gt;</span><span class="ansi-blue-fg">(f)</span>
<span class="ansi-green-intense-fg ansi-bold">    122</span>     bed = _read_file(
<span class="ansi-green-intense-fg ansi-bold">    123</span>         fn<span class="ansi-blue-fg">,</span>
<span class="ansi-green-fg">--&gt; 124</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">lambda</span> f<span class="ansi-blue-fg">:</span> _read_bed<span class="ansi-blue-fg">(</span>f<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;bed&#34;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">,</span> nsamples<span class="ansi-blue-fg">,</span> nmarkers<span class="ansi-blue-fg">[</span>f<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;bed&#34;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">,</span> ref<span class="ansi-blue-fg">,</span> Chunk<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">.</span>T<span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">    125</span>         pbar<span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">    126</span>     )

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/pandas_plink/_read.py</span> in <span class="ansi-cyan-fg">_read_bed</span><span class="ansi-blue-fg">(fn, nsamples, nvariants, ref, chunk)</span>
<span class="ansi-green-intense-fg ansi-bold">    411</span>     col_chunk <span class="ansi-blue-fg">=</span> max<span class="ansi-blue-fg">(</span>ncols <span class="ansi-blue-fg">//</span> max_npartitions<span class="ansi-blue-fg">,</span> col_chunk<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    412</span> 
<span class="ansi-green-fg">--&gt; 413</span><span class="ansi-red-fg">     </span>G <span class="ansi-blue-fg">=</span> read_bed<span class="ansi-blue-fg">(</span>fn<span class="ansi-blue-fg">,</span> nrows<span class="ansi-blue-fg">,</span> ncols<span class="ansi-blue-fg">,</span> row_chunk<span class="ansi-blue-fg">,</span> col_chunk<span class="ansi-blue-fg">,</span> ref<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    414</span>     <span class="ansi-green-fg">if</span> major <span class="ansi-blue-fg">==</span> <span class="ansi-blue-fg">&#34;variant&#34;</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">    415</span>         G <span class="ansi-blue-fg">=</span> G<span class="ansi-blue-fg">.</span>T

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/pandas_plink/_bed_read.py</span> in <span class="ansi-cyan-fg">read_bed</span><span class="ansi-blue-fg">(filepath, nrows, ncols, row_chunk, col_chunk, ref)</span>
<span class="ansi-green-intense-fg ansi-bold">     22</span>     row_size <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">(</span>ncols <span class="ansi-blue-fg">+</span> <span class="ansi-cyan-fg">3</span><span class="ansi-blue-fg">)</span> <span class="ansi-blue-fg">//</span> <span class="ansi-cyan-fg">4</span>
<span class="ansi-green-intense-fg ansi-bold">     23</span>     size <span class="ansi-blue-fg">=</span> nrows <span class="ansi-blue-fg">*</span> row_size
<span class="ansi-green-fg">---&gt; 24</span><span class="ansi-red-fg">     </span>buff <span class="ansi-blue-fg">=</span> memmap<span class="ansi-blue-fg">(</span>filepath<span class="ansi-blue-fg">,</span> uint8<span class="ansi-blue-fg">,</span> <span class="ansi-blue-fg">&#34;r&#34;</span><span class="ansi-blue-fg">,</span> <span class="ansi-cyan-fg">3</span><span class="ansi-blue-fg">,</span> shape<span class="ansi-blue-fg">=</span><span class="ansi-blue-fg">(</span>size<span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     25</span>     row_start <span class="ansi-blue-fg">=</span> <span class="ansi-cyan-fg">0</span>
<span class="ansi-green-intense-fg ansi-bold">     26</span>     col_xs <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">]</span>

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/numpy/core/memmap.py</span> in <span class="ansi-cyan-fg">__new__</span><span class="ansi-blue-fg">(subtype, filename, dtype, mode, offset, shape, order)</span>
<span class="ansi-green-intense-fg ansi-bold">    265</span>             bytes <span class="ansi-blue-fg">-=</span> start
<span class="ansi-green-intense-fg ansi-bold">    266</span>             array_offset <span class="ansi-blue-fg">=</span> offset <span class="ansi-blue-fg">-</span> start
<span class="ansi-green-fg">--&gt; 267</span><span class="ansi-red-fg">             </span>mm <span class="ansi-blue-fg">=</span> mmap<span class="ansi-blue-fg">.</span>mmap<span class="ansi-blue-fg">(</span>fid<span class="ansi-blue-fg">.</span>fileno<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">,</span> bytes<span class="ansi-blue-fg">,</span> access<span class="ansi-blue-fg">=</span>acc<span class="ansi-blue-fg">,</span> offset<span class="ansi-blue-fg">=</span>start<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    268</span> 
<span class="ansi-green-intense-fg ansi-bold">    269</span>             self = ndarray.__new__(subtype, shape, dtype=descr, buffer=mm,

<span class="ansi-red-fg">OSError</span>: [Errno 12] Cannot allocate memory</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/guangyou/imputation/genome/othergenes/UKB_exome_othergenes.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">main</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/guangyou/imputation/genome/othergenes/UKB_exome_othergenes.bed&#39;</span><span class="p">,</span><span class="s1">&#39;test2.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>liftover from hg19tohg38
Removing SNPs failed to liftover is True
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/mnt/mfs/statgen/yin/Github/LDtools/LDtools/genodata.py:83: PerformanceWarning: Slicing is producing a large chunk. To accept the large
chunk and silence this warning, set the option
    &gt;&gt;&gt; with dask.config.set(**{&#39;array.slicing.split_large_chunks&#39;: False}):
    ...     array[indexer]

To avoid creating the large chunks, set the option
    &gt;&gt;&gt; with dask.config.set(**{&#39;array.slicing.split_large_chunks&#39;: True}):
    ...     array[indexer]
  geno = geno[idx,:]
Writing BED: 100%|██████████| 1/1 [00:15&lt;00:00, 15.11s/it]</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Writing FAM... </pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>done.
Writing BIM... done.
Total number SNPs  1413 . Removing SNPs failed to liftover  0
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">main</span><span class="p">(</span><span class="s1">&#39;MWE_region_extraction/100521_UKBB_Hearing_aid_f3393_expandedwhite_15601cases_237318ctrl_500k_PC1_PC2_f3393.regenie.snp_stats.gz&#39;</span><span class="p">,</span><span class="s1">&#39;test_sumstats.sumstats.gz&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>liftover from hg19tohg38
Removing SNPs failed to liftover is True
This file is considered as sumstat format file
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>bim:      chrom                  snp   cm        pos a0 a1     i
0         1    chr1:55039741:G:C  0.0   55039741  G  C     0
1         1    chr1:55039742:G:A  0.0   55039742  G  A     1
2         1    chr1:55039749:G:C  0.0   55039749  G  C     2
3         1    chr1:55039750:T:C  0.0   55039750  T  C     3
4         1    chr1:55039753:T:C  0.0   55039753  T  C     4
...     ...                  ...  ...        ... .. ..   ...
1408     11  chr11:116832956:T:C  0.0  116832956  T  C  1408
1409     11  chr11:116832976:C:G  0.0  116832976  C  G  1409
1410     11  chr11:116832977:T:G  0.0  116832977  T  G  1410
1411     11  chr11:116832978:T:G  0.0  116832978  T  G  1411
1412     11  chr11:116832980:T:C  0.0  116832980  T  C  1412

[1413 rows x 7 columns] 
 fam:            fid      iid father mother gender trait       i
0       1000019  1000019      0      0      2    -9       0
1       1000078  1000078      0      0      2    -9       1
2       1000081  1000081      0      0      1    -9       2
3       1000198  1000198      0      0      2    -9       3
4       1000210  1000210      0      0      1    -9       4
...         ...      ...    ...    ...    ...   ...     ...
168201  6025295  6025295      0      0      1    -9  168201
168202  6025319  6025319      0      0      2    -9  168202
168203  6025346  6025346      0      0      2    -9  168203
168204  6025363  6025363      0      0      1    -9  168204
168205  6025411  6025411      0      0      2    -9  168205

[168206 rows x 7 columns] 
 bed:dask.array&lt;transpose, shape=(1413, 168206), dtype=float32, chunksize=(1024, 1024), chunktype=numpy.ndarray&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno</span><span class="o">.</span><span class="n">extractbyidx</span><span class="p">(</span><span class="o">~</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">chrom</span> <span class="o">==</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno</span><span class="o">.</span><span class="n">export_plink</span><span class="p">(</span><span class="s1">&#39;test1.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>
Writing BED:   0%|          | 0/1 [00:00&lt;?, ?it/s]
Writing BED: 100%|██████████| 1/1 [00:02&lt;00:00,  2.26s/it]</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Writing FAM... </pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>done.
Writing BIM... done.
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Test-liftover-sumstat">Test liftover sumstat<a class="anchor-link" href="#Test-liftover-sumstat"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">LDtools.sumstat</span> <span class="kn">import</span> <span class="n">Sumstat</span>
<span class="kn">from</span> <span class="nn">LDtools.liftover</span> <span class="kn">import</span> <span class="n">Liftover</span>
<span class="k">def</span> <span class="nf">gwas_liftover</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">output_unmapped</span><span class="p">,</span><span class="n">output_mapped</span><span class="p">,</span><span class="n">fr</span><span class="o">=</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span><span class="n">to</span><span class="o">=</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span><span class="n">remove_missing</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="n">lf</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="n">fr</span><span class="p">,</span><span class="n">to</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading GWAS sumstat&quot;</span><span class="p">)</span>
    <span class="n">sums</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;liftover from&quot;</span> <span class="o">+</span> <span class="n">fr</span> <span class="o">+</span><span class="s2">&quot;to&quot;</span> <span class="o">+</span><span class="n">to</span><span class="p">)</span>
    <span class="n">sums1</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">sumstat_liftover</span><span class="p">(</span><span class="n">sums</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">remove_missing</span><span class="p">:</span>
        <span class="n">sums1</span><span class="p">[</span><span class="n">sums1</span><span class="o">.</span><span class="n">CHR</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_unmapped</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">sums1</span><span class="p">[</span><span class="n">sums1</span><span class="o">.</span><span class="n">CHR</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_mapped</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sums1</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">gwas_liftover</span><span class="p">(</span><span class="n">input_file</span><span class="p">,</span><span class="n">output_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">fr</span><span class="o">=</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span><span class="n">to</span><span class="o">=</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span><span class="n">remove_missing</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">output_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;/&#39;</span>
    <span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
    <span class="n">lf</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span><span class="s1">&#39;hg38&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reading GWAS sumstat&quot;</span><span class="p">)</span>
    <span class="n">sums</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;liftover from&quot;</span> <span class="o">+</span> <span class="n">fr</span> <span class="o">+</span><span class="s2">&quot;to&quot;</span> <span class="o">+</span><span class="n">to</span><span class="p">)</span>
    <span class="n">sums1</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">sumstat_liftover</span><span class="p">(</span><span class="n">sums</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">remove_missing</span><span class="p">:</span>
        <span class="n">sums1</span><span class="p">[</span><span class="n">sums1</span><span class="o">.</span><span class="n">CHR</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_unmapped</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">sums1</span><span class="p">[</span><span class="n">sums1</span><span class="o">.</span><span class="n">CHR</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_mapped</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sums1</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sumstats_lifted</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.hg38.sumstats.gz&#39;</span><span class="p">,</span>
<span class="n">sumstats_unmapped</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.hg38.sumstats_unmapped.gz&#39;</span><span class="p">,</span>
<span class="n">sumstats_mapped</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.hg38.sumstats_mapped.gz&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">tmp</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(&#39;100521_UKBB_Combined_f2247_f2257_expandedwhite_93258cases_237318ctrl_500k_PC1_PC2_f2247_f2257.regenie.snp_stats&#39;,
 &#39;.gz&#39;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;/&#39;</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;/home/dmc2245/UKBiobank/results/REGENIE_results/results_imputed_data/2021_10_07_combined_500K/&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sums</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span><span class="s1">&#39;hg38&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">sumstat_liftover</span><span class="p">(</span><span class="n">sums</span><span class="o">.</span><span class="n">ss</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CHR</th>
      <th>POS</th>
      <th>REF</th>
      <th>ALT</th>
      <th>SNP</th>
      <th>BETA</th>
      <th>SE</th>
      <th>P</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>13259</td>
      <td>G</td>
      <td>A</td>
      <td>chr1:13259:G:A</td>
      <td>0.434586</td>
      <td>0.175780</td>
      <td>0.014801</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>17569</td>
      <td>C</td>
      <td>A</td>
      <td>chr1:17569:C:A</td>
      <td>-0.030568</td>
      <td>0.795968</td>
      <td>0.969366</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>17641</td>
      <td>G</td>
      <td>A</td>
      <td>chr1:17641:G:A</td>
      <td>-0.078881</td>
      <td>0.108663</td>
      <td>0.467883</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>30741</td>
      <td>C</td>
      <td>A</td>
      <td>chr1:30741:C:A</td>
      <td>-1.599610</td>
      <td>0.990472</td>
      <td>0.044798</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1</td>
      <td>57222</td>
      <td>T</td>
      <td>C</td>
      <td>chr1:57222:T:C</td>
      <td>0.031666</td>
      <td>0.121422</td>
      <td>0.794253</td>
    </tr>
    <tr>
      <th>5</th>
      <td>1</td>
      <td>58396</td>
      <td>T</td>
      <td>C</td>
      <td>chr1:58396:T:C</td>
      <td>0.366266</td>
      <td>0.172004</td>
      <td>0.035663</td>
    </tr>
    <tr>
      <th>6</th>
      <td>1</td>
      <td>62157</td>
      <td>G</td>
      <td>A</td>
      <td>chr1:62157:G:A</td>
      <td>-0.147251</td>
      <td>0.296105</td>
      <td>0.618983</td>
    </tr>
    <tr>
      <th>7</th>
      <td>1</td>
      <td>62595</td>
      <td>C</td>
      <td>T</td>
      <td>chr1:62595:C:T</td>
      <td>0.356993</td>
      <td>0.171623</td>
      <td>0.040096</td>
    </tr>
    <tr>
      <th>8</th>
      <td>1</td>
      <td>69487</td>
      <td>G</td>
      <td>A</td>
      <td>chr1:69487:G:A</td>
      <td>-0.559373</td>
      <td>0.853882</td>
      <td>0.512407</td>
    </tr>
    <tr>
      <th>9</th>
      <td>1</td>
      <td>69569</td>
      <td>T</td>
      <td>C</td>
      <td>chr1:69569:T:C</td>
      <td>0.232690</td>
      <td>0.216585</td>
      <td>0.282662</td>
    </tr>
  </tbody>
</table>
</div></div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">remove_missing</span><span class="p">):</span>
    <span class="n">sums</span> <span class="o">=</span> <span class="n">read_regenie</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">sums1</span> <span class="o">=</span> <span class="n">sumstat_liftover</span><span class="p">(</span><span class="n">sums</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">remove_missing</span><span class="p">:</span>
        <span class="n">sums1</span><span class="p">[</span><span class="n">sums1</span><span class="o">.</span><span class="n">CHR</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sums1</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">input_path</span> <span class="o">=</span> <span class="s1">&#39;/home/dmc2245/UKBiobank/results/REGENIE_results/results_imputed_data/2021_10_07_combined_500K/100521_UKBB_Combined_f2247_f2257_expandedwhite_93258cases_237318ctrl_500k_PC1_PC2_f2247_f2257.regenie.snp_stats.gz&#39;</span>
<span class="n">output_path</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="n">remove_missing</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">main</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span><span class="n">output_path</span><span class="p">,</span><span class="n">remove_missing</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="s1">&#39;:&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="s1">&#39;1&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">TypeError</span>                                 Traceback (most recent call last)
<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_39087/731612175.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">----&gt; 1</span><span class="ansi-red-fg"> </span><span class="ansi-blue-fg">&#39;:&#39;</span><span class="ansi-blue-fg">.</span>join<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">[</span><span class="ansi-cyan-fg">1</span><span class="ansi-blue-fg">,</span><span class="ansi-blue-fg">&#39;1&#39;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">)</span>

<span class="ansi-red-fg">TypeError</span>: sequence item 0: expected str instance, int found</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gen</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="test-chr22">test chr22<a class="anchor-link" href="#test-chr22"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">region</span> <span class="o">=</span> <span class="p">[</span><span class="mi">22</span><span class="p">,</span><span class="mi">50519304</span><span class="p">,</span><span class="mi">50549676</span><span class="p">]</span>
<span class="n">exome_sumstats</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="s1">&#39;/home/dmc2245/UKBiobank/results/REGENIE_results/results_imputed_data/010522_f2247_hearing_diff_200K_imputed/090321_UKBB_Hearing_difficulty_f2247_expandedwhite_45502cases_96601ctrl_PC1_2_f2247.regenie.snp_stats.gz&#39;</span><span class="p">)</span>
<span class="n">exome_sumstats</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>this region [22, 50519304, 50549676] has 669 SNPs in Sumstat
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1.2. LiftOver the region&#39;</span><span class="p">)</span>
<span class="n">hg38toimpref</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span><span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
<span class="n">imp_region</span> <span class="o">=</span> <span class="n">hg38toimpref</span><span class="o">.</span><span class="n">region_liftover</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
<span class="n">imput_sumstats</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">imp_region</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1.3. Regional SNPs Liftover&#39;</span><span class="p">)</span>
<span class="n">impreftohg38</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="n">imp_ref</span><span class="p">,</span><span class="s1">&#39;hg38&#39;</span><span class="p">)</span> <span class="c1">#oppsite with hg38toimpref</span>
<span class="n">impreftohg38</span><span class="o">.</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

